File: plugins/plugin-shell/typescript/actions/clearHistory.ts
import {
  type Action,
  type ActionExample,
  type Content,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { ShellService } from "../services/shellService";

const spec = requireActionSpec("CLEAR_HISTORY");

export const clearHistory: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (runtime: IAgentRuntime, message: Memory, _state: State): Promise<boolean> => {
    const shellService = runtime.getService<ShellService>("shell");
    if (!shellService) {
      return false;
    }

    const text = message.content.text?.toLowerCase() || "";
    const clearKeywords = ["clear", "reset", "delete", "remove", "clean"];
    const historyKeywords = ["history", "terminal", "shell", "command"];

    const hasClearKeyword = clearKeywords.some((keyword) => text.includes(keyword));
    const hasHistoryKeyword = historyKeywords.some((keyword) => text.includes(keyword));

    return hasClearKeyword && hasHistoryKeyword;
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ) => {
    const shellService = runtime.getService<ShellService>("shell");

    if (!shellService) {
      if (callback) {
        await callback({
          text: "Shell service is not available.",
          source: message.content.source,
        });
      }
      return { success: false, error: "Shell service is not available." };
    }

    const conversationId = message.roomId || message.agentId;
    if (!conversationId) {
      const errorMsg = "No conversation ID available";
      if (callback) {
        await callback({
          text: errorMsg,
          source: message.content.source,
        });
      }
      return { success: false, error: errorMsg };
    }
    shellService.clearCommandHistory(conversationId);

    logger.info(`Cleared shell history for conversation: ${conversationId}`);

    const response: Content = {
      text: "Shell command history has been cleared.",
      source: message.content.source,
    };

    if (callback) {
      await callback(response);
    }
    return { success: true, text: response.text };
  },
  examples: (spec.examples ?? []) as ActionExample[][],
};

export default clearHistory;

File: plugins/plugin-shell/rust/src/actions/clear_history.rs
use crate::{Action, ActionExample, ActionResult, ShellService};
use async_trait::async_trait;
use serde_json::Value;

pub struct ClearHistoryAction;

impl ClearHistoryAction {
    const CLEAR_KEYWORDS: &'static [&'static str] =
        &["clear", "reset", "delete", "remove", "clean"];
    const HISTORY_KEYWORDS: &'static [&'static str] = &["history", "terminal", "shell", "command"];

    fn has_clear_keyword(text: &str) -> bool {
        let lower = text.to_lowercase();
        Self::CLEAR_KEYWORDS.iter().any(|kw| lower.contains(kw))
    }

    fn has_history_keyword(text: &str) -> bool {
        let lower = text.to_lowercase();
        Self::HISTORY_KEYWORDS.iter().any(|kw| lower.contains(kw))
    }
}

#[async_trait]
impl Action for ClearHistoryAction {
    fn name(&self) -> &str {
        "CLEAR_SHELL_HISTORY"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "RESET_SHELL",
            "CLEAR_TERMINAL",
            "CLEAR_HISTORY",
            "RESET_HISTORY",
        ]
    }

    fn description(&self) -> &str {
        "Clears the recorded history of shell commands for the current conversation"
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        Self::has_clear_keyword(text) && Self::has_history_keyword(text)
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut ShellService>,
    ) -> ActionResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Shell service is not available.".to_string(),
                    data: None,
                    error: Some("Shell service is not available".to_string()),
                }
            }
        };

        let conversation_id = message
            .get("room_id")
            .and_then(|r| r.as_str())
            .or_else(|| message.get("agent_id").and_then(|a| a.as_str()))
            .unwrap_or("default");

        service.clear_command_history(conversation_id);

        ActionResult {
            success: true,
            text: "Shell command history has been cleared.".to_string(),
            data: None,
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "clear my shell history".to_string(),
                agent_response: "Shell command history has been cleared.".to_string(),
            },
            ActionExample {
                user_message: "reset the terminal history".to_string(),
                agent_response: "Shell command history has been cleared.".to_string(),
            },
            ActionExample {
                user_message: "delete command history".to_string(),
                agent_response: "Shell command history has been cleared.".to_string(),
            },
        ]
    }
}
